src/feature_utils_test.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2021 Google LLC
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     https://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for feature utils."""
+
+from absl.testing import parameterized
+import tensorflow as tf
+
+import feature_utils
+
+
+class TensorUtilsTest(tf.test.TestCase, parameterized.TestCase):
+
+  def test_relative_position_generator_init(self):
+    relative_pos_gen = feature_utils.MmtRelativePositionGenerator(
+        num_patch_per_row=2,
+        num_core_layers=1,
+        text_relative_pos_max_distance=3)
+
+    self.assertEqual(2, relative_pos_gen._num_patch_per_row)
+    self.assertEqual(1, relative_pos_gen._num_core_layers)
+    self.assertEqual(3, relative_pos_gen._core_layer_diameter)
+    self.assertEqual(19, relative_pos_gen._image_part_id)
+    self.assertEqual(20, relative_pos_gen._text_part_id)
+
+  def test_relative_position_generator_init_invalid_arguments(self):
+    with self.assertRaises(ValueError):
+      feature_utils.MmtRelativePositionGenerator(num_patch_per_row=0,
+                                                 num_core_layers=1,
+                                                 text_relative_pos_max_distance=2)
+      feature_utils.MmtRelativePositionGenerator(num_patch_per_row=1,
+                                                 num_core_layers=0,
+                                                 text_relative_pos_max_distance=2)
+      feature_utils.MmtRelativePositionGenerator(num_patch_per_row=1,
+                                                 num_core_layers=1,
+                                                 text_relative_pos_max_distance=-1)
+
+  def test_make_relative_att_ids_smaller_case(self):
+    relative_pos_gen = feature_utils.MmtRelativePositionGenerator(
+        num_patch_per_row=2,
+        num_core_layers=1,
+        text_relative_pos_max_distance=3)
+
+    """
+    base_tensor = tf.Tensor(
+        [[16  9  9  9 10]
+         [15  5  6  7 11]
+         [15  8  0  1 11]
+         [15  2  3  4 11]
+         [14 13 13 13 12]], shape=(5, 5), dtype=int32)
+    """
+
+    expected = [[
```

Comment:
Probably the problem is the "num_patch_per_row" and "num_core_layers" parameters, which I still do not understand. But I do not understand this expected output and what each row/column means.

Reply:
I updated docstring to explain.